File: Python/OneLang/One/Transforms/LambdaCaptureCollector.py
from onelang_core import *
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Interfaces as ints
import OneLang.One.Ast.References as refs

class LambdaCaptureCollector(astTrans.AstTransformer):
    def __init__(self):
        self.scope_var_stack = []
        self.scope_vars = None
        self.captured_vars = None
        super().__init__("LambdaCaptureCollector")
    
    def visit_lambda(self, lambda_):
        if self.scope_vars != None:
            self.scope_var_stack.append(self.scope_vars)
        
        self.scope_vars = dict()
        self.captured_vars = dict()
        
        super().visit_lambda(lambda_)
        lambda_.captures = []
        for capture in self.captured_vars.keys():
            lambda_.captures.append(capture)
        
        self.scope_vars = self.scope_var_stack.pop() if len(self.scope_var_stack) > 0 else None
        return lambda_
    
    def visit_variable(self, variable):
        if self.scope_vars != None:
            self.scope_vars[variable] = None
        return variable
    
    def visit_variable_reference(self, var_ref):
        if isinstance(var_ref, refs.StaticFieldReference) or isinstance(var_ref, refs.InstanceFieldReference) or isinstance(var_ref, refs.StaticPropertyReference) or isinstance(var_ref, refs.InstancePropertyReference) or self.scope_vars == None:
            return var_ref
        
        vari = var_ref.get_variable()
        if not vari in self.scope_vars:
            self.captured_vars[vari] = None
        
        return var_ref

File: PHP/src/OneLang/One/Transforms/LambdaCaptureCollector.php
<?php

namespace OneLang\One\Transforms\LambdaCaptureCollector;

use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\Ast\AstTypes\UnresolvedType;
use OneLang\One\Ast\AstTypes\GenericsType;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\IVariable;
use OneLang\One\Ast\Types\Lambda;
use OneLang\One\Ast\Types\Method;
use OneLang\One\Ast\Interfaces\IType;
use OneLang\One\Ast\References\InstanceFieldReference;
use OneLang\One\Ast\References\InstancePropertyReference;
use OneLang\One\Ast\References\StaticFieldReference;
use OneLang\One\Ast\References\StaticPropertyReference;
use OneLang\One\Ast\References\VariableReference;

class LambdaCaptureCollector extends AstTransformer {
    public $scopeVarStack;
    public $scopeVars;
    public $capturedVars;
    
    function __construct() {
        parent::__construct("LambdaCaptureCollector");
        $this->scopeVarStack = array();
        $this->scopeVars = null;
        $this->capturedVars = null;
    }
    
    protected function visitLambda($lambda) {
        if ($this->scopeVars !== null)
            $this->scopeVarStack[] = $this->scopeVars;
        
        $this->scopeVars = new \OneLang\Core\Set();
        $this->capturedVars = new \OneLang\Core\Set();
        
        parent::visitLambda($lambda);
        $lambda->captures = array();
        foreach ($this->capturedVars->values() as $capture)
            $lambda->captures[] = $capture;
        
        $this->scopeVars = count($this->scopeVarStack) > 0 ? array_pop($this->scopeVarStack) : null;
        return $lambda;
    }
    
    protected function visitVariable($variable) {
        if ($this->scopeVars !== null)
            $this->scopeVars->add($variable);
        return $variable;
    }
    
    protected function visitVariableReference($varRef) {
        if ($varRef instanceof StaticFieldReference || $varRef instanceof InstanceFieldReference || $varRef instanceof StaticPropertyReference || $varRef instanceof InstancePropertyReference || $this->scopeVars === null)
            return $varRef;
        
        $vari = $varRef->getVariable();
        if (!$this->scopeVars->has($vari))
            $this->capturedVars->add($vari);
        
        return $varRef;
    }
}
